src/Model/DriverOptions.php

```diff
@@ -0,0 +1,163 @@
+                'typeMap' => $this->typeMap,
+                'builderEncoder' => $this->builderEncoder,
+                'autoEncryption' => $this->autoEncryption,
+                'driver' => $this->driver,
+            ] + $this->miscOptions,
+            fn ($option) => $option !== null,
+        );
+    }
+
+    /** @throws InvalidArgumentException */
+    private static function ensureValidArrayOptions(array $options): void
+    {
+        if (! is_array($options[self::KEY_TYPE_MAP])) {
+            throw InvalidArgumentException::invalidType(
+                sprintf('"%s" driver option', self::KEY_TYPE_MAP),
+                $options[self::KEY_TYPE_MAP],
+                'array',
+            );
+        }
+
+        if (isset($options[self::KEY_BUILDER_ENCODER]) && ! $options[self::KEY_BUILDER_ENCODER] instanceof Encoder) {
+            throw InvalidArgumentException::invalidType(
+                sprintf('"%s" option', self::KEY_BUILDER_ENCODER),
+                $options[self::KEY_BUILDER_ENCODER],
+                Encoder::class,
+            );
+        }
+
+        $driver = $options[self::KEY_DRIVER] ?? [];
+        if (isset($driver['name'])) {
+            if (! is_string($driver['name'])) {
+                throw InvalidArgumentException::invalidType('"name" handshake option', $driver['name'], 'string');
+            }
+        }
+
+        if (isset($driver['version'])) {
+            if (! is_string($driver['version'])) {
+                throw InvalidArgumentException::invalidType('"version" handshake option', $driver['version'], 'string');
+            }
+        }
+    }
+
+    private static function getVersion(): string
+    {
+        if (self::$version === null) {
+            try {
+                self::$version = InstalledVersions::getPrettyVersion('mongodb/mongodb') ?? 'unknown';
+            } catch (Throwable) {
+                self::$version = 'error';
+            }
+        }
+
+        return self::$version;
+    }
+
+    private function withDriverInfo(array $driver): self
+    {
+        $mergedDriver = [
+            'name' => 'PHPLIB',
+            'version' => self::getVersion(),
+        ];
+
+        if (isset($driver['name'])) {
+            $mergedDriver['name'] .= self::HANDSHAKE_SEPARATOR . $driver['name'];
+        }
+
+        if (isset($driver['version'])) {
+            $mergedDriver['version'] .= self::HANDSHAKE_SEPARATOR . $driver['version'];
+        }
+
+        if (isset($driver['platform'])) {
+            $mergedDriver['platform'] = $driver['platform'];
+        }
+
+        if ($this->isAutoEncryptionEnabled()) {
+            $mergedDriver['platform'] = trim(sprintf('iue %s', $driver['platform'] ?? ''));
+        }
+
+        return new self(
+            typeMap: $this->typeMap,
+            builderEncoder: $this->builderEncoder,
+            autoEncryption: $this->autoEncryption,
+            driver: $mergedDriver,
+            miscOptions: $this->miscOptions,
+        );
+    }
```

Comment:
Creating a second instance of `DriverOptions` does not seem efficient to me, especially since this method is always called in the factory. A simple assignment  is enough.
```suggestion
        $this->driver = $mergedDriver;
    }
```

Reply:
The thing is that then I'd have to change the `readonly` and visibility (which is fine, I can make it private I guess. Nothing is directly accessing it as of now)

Reply:
If you don't use CPP for `$driver`, you can alter its value before setting the property.